core/lib/src/fs/maybe_zipped_file.rs

```diff
@@ -0,0 +1,110 @@
+use std::ffi::OsString;
+use std::path::Path;
+use std::io;
+
+use rocket_http::ContentType;
+
+use crate::fs::NamedFile;
+use crate::response::{self, Responder};
+use crate::Request;
+
+/// A [`Responder`] that looks for pre-zipped files on the filesystem (by an 
+/// extra `.gz' file extension) to serve in place of the given file.
+///
+/// # Example
+///
+/// A simple static file server mimicking [`FileServer`]:
+///
+/// ```rust
+/// # use rocket::get;
+/// use std::path::{PathBuf, Path};
+///
+/// use rocket::fs::{NamedFile, relative};
+///
+/// #[get("/file/<path..>")]
+/// pub async fn second(path: PathBuf) -> Option<NamedFile> {
+///     let mut path = Path::new(relative!("static")).join(path);
+///     if path.is_dir() {
+///         path.push("index.html");
+///     }
+///
+///     MaybeZippedFile::open(path).await.ok()
+/// }
+/// ```
+///
+/// Always prefer to use [`FileServer`] which has more functionality and a
+/// pithier API.
+///
+/// [`FileServer`]: crate::fs::FileServer
+#[derive(Debug)]
+pub(crate) struct MaybeZippedFile {
+    ct_ext: Option<OsString>,
+    encoded: bool,
+    file: NamedFile, 
+}
+
+impl MaybeZippedFile {
+    /// Attempts to open files in read-only mode.
+    /// 
+    /// # Errors
+    ///
+    /// This function will return an error if the selected path does not already 
+    /// exist. Other errors may also be returned according to
+    /// [`OpenOptions::open()`](std::fs::OpenOptions::open()).
+    pub async fn open<P: AsRef<Path>>(path: P) -> io::Result<MaybeZippedFile> {
+        let o_path = path.as_ref().to_path_buf();
+
+        let (ct_ext, encoded, file) = match o_path.extension() {
+            // A gz file is being requested, no need to compress again.
+            Some(e) if e == "gz" => (Some(e.to_owned()), false, NamedFile::open(path).await?),
+            // construct path to the .gz file
+            Some(e) => {
+                let bare = o_path.with_extension("");
+
+                let ct_ext = Some(e.to_owned());
+                let mut zip_ext = e.to_owned();
+                zip_ext.push(".gz");
+
+                let zipped = bare.with_extension(&zip_ext);
```

Comment:
I think there is a bug here. If the requested file is something like `index.something.html`, then `bare` will become `index.something`. Calling `with_extension("html.gz)` on this will remove the file extension once again and result in `index.html.gz`.

I think that you can just call `o_path.with_extension(&zip_ext)` here, `bare` is unnecessary.

Reply:
Oh! Good catch. I'll give that a whirl.